runtime/resolvers/resolvers_test.go

```diff
@@ -0,0 +1,124 @@
+	"gopkg.in/yaml.v3"
+)
+
+type Project struct {
+	Sources    map[string]yaml.Node
+	Models     map[string]yaml.Node
+	Dashboards map[string]yaml.Node
+	APIs       map[string]yaml.Node
+}
+type Resolvers struct {
+	Project    Project
+	Connectors map[string]testruntime.InstanceOptions
+	Tests      map[string]*Test
+}
+
+type Test struct {
+	Resolver string
+	Options  runtime.ResolveOptions
+	Result   []map[string]any
+}
+
+func TestResolvers(t *testing.T) {
+	yamlFile, err := os.ReadFile("resolvers_test.yaml")
+	require.NoError(t, err)
+	var r Resolvers
+	err = yaml.Unmarshal(yamlFile, &r)
+	require.NoError(t, err)
+
+	files := make(map[string]string)
+	for name, node := range r.Project.Sources {
+		abs := filepath.Join("sources", name)
+		bytes, err := yaml.Marshal(&node)
+		require.NoError(t, err)
+		files[abs] = string(bytes)
+
+		// copy the data file to the temporary folder as well
+		var props map[string]string
+		require.NoError(t, node.Decode(&props))
+		if props["connector"] == "local_file" {
+			abs := props["path"]
+			bs, err := os.ReadFile(props["path"])
+			require.NoError(t, err)
+			files[abs] = string(bs)
+		}
+	}
+	for name, node := range r.Project.Models {
+		abs := filepath.Join("models", name)
+		var bytes []byte
+		// reading SQL without '|' from YAML
+		if strings.HasSuffix(name, "sql") {
+			bytes = []byte(node.Value)
+		} else {
+			bytes, err = yaml.Marshal(&node)
+			require.NoError(t, err)
+		}
+		files[abs] = string(bytes)
+
+	}
+	for name, node := range r.Project.Dashboards {
+		abs := filepath.Join("dashboards", name)
+		bytes, err := yaml.Marshal(&node)
+		require.NoError(t, err)
+		files[abs] = string(bytes)
+	}
+	for name, node := range r.Project.APIs {
+		abs := filepath.Join("apis", name)
+		bytes, err := yaml.Marshal(&node)
+		require.NoError(t, err)
+		files[abs] = string(bytes)
+
+	}
+
+	for ct, opts := range r.Connectors {
+		switch ct {
+		case "duckdb":
+			fmt.Println("Running with DuckDB")
+			if opts.Files == nil {
+				opts.Files = map[string]string{"rill.yaml": ""}
+			}
+			maps.Copy(opts.Files, files)
+			rt, instanceID := testruntime.NewInstanceWithOptions(t, opts)
+			for testName, test := range r.Tests {
+				t.Run(testName, func(t *testing.T) {
+					t.Log("======================")
+					t.Log("Running ", testName)
+					testruntime.RequireParseErrors(t, rt, instanceID, nil)
+					api, err := rt.APIForName(context.Background(), instanceID, test.Resolver)
+					require.NoError(t, err)
+
+					o := test.Options
+					o.InstanceID = instanceID
+					o.Resolver = api.Spec.Resolver
+					o.ResolverProperties = api.Spec.ResolverProperties.AsMap()
+
+					res, err := rt.Resolve(context.Background(), &o)
+					require.NoError(t, err)
+					var rows []map[string]interface{}
+					require.NoError(t, json.Unmarshal(res.Data, &rows), string(res.Data))
+					assert := td.Assert(t)
+					if !assert.Cmp(rows, test.Result) {
+						t.Log("Failure") // breakpoint can be placed here
+					}
```

Comment:
Can we use the `testify/require` package we use elsewhere? It also has deep comparison functions

Reply:
I wonder why replacing `assert` - it allows deep comparison of structures.

Reply:
Done.